pkg/pf/tests/upgrade_state_cross_test.go

```diff
@@ -0,0 +1,1133 @@
+			return map[int64]resource.StateUpgrader{
+				0: {
+					PriorSchema: &resourceBefore.ResourceSchema,
+					StateUpgrader: func(
+						ctx context.Context,
+						req resource.UpgradeStateRequest,
+						resp *resource.UpgradeStateResponse,
+					) {
+						autogold.Expect(`{
+            "id": "test-id",
+            "prop": "one,two,three"
+          }`).Equal(t, string(req.RawState.JSON))
+
+						var priorState v0Model
+						d0 := req.State.Get(ctx, &priorState)
+						resp.Diagnostics = append(resp.Diagnostics, d0...)
+
+						partsN := []int32{}
+						parts := strings.Split(priorState.Prop, ",")
+						for _, p := range parts {
+							parsed := map[string]int32{"one": 1, "two": 2, "three": 3}
+							partsN = append(partsN, parsed[p])
+						}
+
+						updState := v1Model{
+							ID:   priorState.ID,
+							Prop: partsN,
+						}
+
+						d := resp.State.Set(ctx, updState)
+						resp.Diagnostics = append(resp.Diagnostics, d...)
+
+						if d.HasError() {
+							panic(fmt.Sprintf("D!: %v", resp.Diagnostics))
+						}
+					},
+				},
+			}
+		},
+	})
+
+	result := runUpgradeStateTest(t, upgradeStateTestCase{
+		Resource1:  &resourceBefore,
+		Resource2:  &resourceAfter,
+		Inputs1:    cty.ObjectVal(map[string]cty.Value{"prop": cty.StringVal("one,two,three")}),
+		InputsMap1: presource.PropertyMap{"prop": presource.NewStringProperty("one,two,three")},
+		Inputs2: cty.ObjectVal(map[string]cty.Value{"prop": cty.ListVal([]cty.Value{
+			cty.NumberIntVal(1),
+			cty.NumberIntVal(2),
+			cty.NumberIntVal(3),
+		})}),
+		InputsMap2: presource.PropertyMap{"prop": presource.NewArrayProperty([]presource.PropertyValue{
+			presource.NewNumberProperty(1),
+			presource.NewNumberProperty(2),
+			presource.NewNumberProperty(3),
+		})},
+
+		// Apparently in this case TF expects RawState to be received on the new schema.
+		ExpectedRawStateType: resourceAfter.ResourceSchema.Type().TerraformType(context.Background()),
+
+		SkipPulumi: "TODO[pulumi/pulumi-terraform-bridge#1667] raw state does not parse properly",
+	})
+
+	autogold.Expect([]upgradeStateTrace{}).Equal(t, result.pulumiUpgrades)
+	autogold.Expect([]upgradeStateTrace{
+		{
+			Phase: upgradeStateTestPhase("refresh"),
+			PriorState: map[string]interface{}{
+				"id":   "test-id",
+				"prop": "one,two,three",
+			},
+			ReturnedState: map[string]interface{}{
+				"id": "test-id",
+				"prop": []interface{}{
+					1,
+					2,
+					3,
+				},
+			},
+		},
+		{
+			Phase: upgradeStateTestPhase("preview"),
+			PriorState: map[string]interface{}{
+				"id":   "test-id",
+				"prop": "one,two,three",
+			},
+			ReturnedState: map[string]interface{}{
+				"id": "test-id",
+				"prop": []interface{}{
+					1,
+					2,
+					3,
+				},
+			},
+		},
+	}).Equal(t, result.tfUpgrades)
+}
+
+// Pulumi removing MaxItems=1 without TF schema changes should be tolerated, without calling upgraders.
+func TestPFUpgrade_Pulumi_Removes_MaxItems1(t *testing.T) {
```

Comment:
Noting that this needs a meaningful refresh implementation, possibly like most of these tests.